In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#### Loading the original data
The dataset consists of 3 csv files -
 * **goalscorers.csv**, which contains all the goals scores in international football history
 * **results.csv**, which contains all the matches since 1876
 * **shootouts.csv** which contains all the penalty shoot-outs.

In [2]:
goals = pd.read_csv('data/goalscorers.csv')
matches = pd.read_csv('data/results.csv')
penalties = pd.read_csv('data/shootouts.csv')

Let us have a quick look at the data.

In [3]:
goals

date  home_team away_team       team            scorer  minute  \
0      1916-07-02      Chile   Uruguay    Uruguay   José Piendibene    44.0   
1      1916-07-02      Chile   Uruguay    Uruguay  Isabelino Gradín    55.0   
2      1916-07-02      Chile   Uruguay    Uruguay  Isabelino Gradín    70.0   
3      1916-07-02      Chile   Uruguay    Uruguay   José Piendibene    75.0   
4      1916-07-06  Argentina     Chile  Argentina     Alberto Ohaco     2.0   
...           ...        ...       ...        ...               ...     ...   
41003  2022-12-18  Argentina    France  Argentina    Ángel Di María    36.0   
41004  2022-12-18  Argentina    France     France     Kylian Mbappé    80.0   
41005  2022-12-18  Argentina    France     France     Kylian Mbappé    81.0   
41006  2022-12-18  Argentina    France  Argentina      Lionel Messi   109.0   
41007  2022-12-18  Argentina    France     France     Kylian Mbappé   118.0   

       own_goal  penalty  
0         False    False  
1         False    False  
2         False    False  
3         False    False  
4         False    False  
...         ...      ...  
41003     False    False  
41004     False     True  
41005     False    False  
41006     False    False  
41007     False     True  

[41008 rows x 8 columns]

In [4]:
matches

date                     home_team                 away_team  \
0      1872-11-30                      Scotland                   England   
1      1873-03-08                       England                  Scotland   
2      1874-03-07                      Scotland                   England   
3      1875-03-06                       England                  Scotland   
4      1876-03-04                      Scotland                   England   
...           ...                           ...                       ...   
44552  2023-03-28                         Haiti                   Bermuda   
44553  2023-03-28                        Guyana                Montserrat   
44554  2023-03-28                       Bonaire  Turks and Caicos Islands   
44555  2023-03-28  United States Virgin Islands              Sint Maarten   
44556  2023-03-29                        Rwanda                     Benin   

       home_score  away_score                            tournament  \
0               0           0                              Friendly   
1               4           2                              Friendly   
2               2           1                              Friendly   
3               2           2                              Friendly   
4               3           0                              Friendly   
...           ...         ...                                   ...   
44552           3           1               CONCACAF Nations League   
44553           0           0               CONCACAF Nations League   
44554           1           2               CONCACAF Nations League   
44555           1           2               CONCACAF Nations League   
44556           1           1  African Cup of Nations qualification   

                city                       country  neutral  
0            Glasgow                      Scotland    False  
1             London                       England    False  
2            Glasgow                      Scotland    False  
3             London                       England    False  
4            Glasgow                      Scotland    False  
...              ...                           ...      ...  
44552  San Cristóbal            Dominican Republic     True  
44553     Bridgetown                      Barbados     True  
44554         Rincon                       Bonaire    False  
44555    Saint Croix  United States Virgin Islands    False  
44556         Kigali                        Rwanda    False  

[44557 rows x 9 columns]

In [5]:
penalties

date    home_team         away_team       winner
0    1967-08-22        India            Taiwan       Taiwan
1    1971-11-14  South Korea  Vietnam Republic  South Korea
2    1972-05-07  South Korea              Iraq         Iraq
3    1972-05-17     Thailand       South Korea  South Korea
4    1972-05-19     Thailand          Cambodia     Thailand
..          ...          ...               ...          ...
542  2022-12-05        Japan           Croatia      Croatia
543  2022-12-06      Morocco             Spain      Morocco
544  2022-12-09      Croatia            Brazil      Croatia
545  2022-12-09  Netherlands         Argentina    Argentina
546  2022-12-18    Argentina            France    Argentina

[547 rows x 4 columns]

Looking at the first two tables, I feel one more table can be formed by merging the two, which can be quite useful because then we will have all the goals in every match segregated at a place, instead of two different tables for the same purpose.

However, merging operations might lead to increase the number of missing values even further if there already are missing values in the dataframes. So, we need to check for and treat missing values before merging.

In [21]:
matches.isna().sum()    #counts missing values for each column

date          0
home_team     0
away_team     0
home_score    0
away_score    0
tournament    0
city          0
country       0
neutral       0
dtype: int64

In [22]:
penalties.isna().sum()

date         0
home_team    0
away_team    0
winner       0
dtype: int64

Thankfully, **matches** and **penalties** tables do not have any missing values.

In [23]:
goals.isna().sum()

date           0
home_team      0
away_team      0
team           0
scorer        49
minute       258
own_goal       0
penalty        0
dtype: int64

In [26]:
goals[['scorer', 'minute']][goals[['scorer', 'minute']]['scorer'].isna()]   #to check if both columns intersect at missing values

scorer  minute
8012    NaN     NaN
8013    NaN     NaN
8014    NaN     NaN
8015    NaN     NaN
8016    NaN     NaN
8017    NaN     NaN
8018    NaN     NaN
8019    NaN     NaN
8020    NaN     NaN
8021    NaN     NaN
8022    NaN     NaN
8040    NaN     NaN
8041    NaN     NaN
8042    NaN     NaN
8043    NaN     NaN
8044    NaN     NaN
8045    NaN     NaN
8046    NaN     NaN
8059    NaN     NaN
8060    NaN     NaN
8061    NaN     NaN
8062    NaN     NaN
8063    NaN     NaN
8064    NaN     NaN
8065    NaN     NaN
8066    NaN     NaN
8067    NaN     NaN
8068    NaN     NaN
8069    NaN     NaN
8070    NaN     NaN
8071    NaN     NaN
8073    NaN     NaN
8074    NaN     NaN
8075    NaN     NaN
8076    NaN     NaN
8077    NaN     NaN
8078    NaN     NaN
8079    NaN     NaN
8080    NaN     NaN
8081    NaN     NaN
8082    NaN     NaN
8083    NaN     NaN
8084    NaN     NaN
8085    NaN     NaN
8086    NaN     NaN
8087    NaN     NaN
8088    NaN     NaN
8089    NaN     NaN
8293    NaN     NaN

Clearly, the <i>scorer</i> and <i>minute</i> columns have missing values, whenever <i>scorer</i> is missing, <i>minute</i> also is missing.
For once, we can go ahead and remove the respective rows but, I would want to lose as little data as possible because the dataset is already pretty small (just above 40k observations).

So, here is what I decided to do -
* Remove the rows with missing <i>scorer</i>, since these form merely 0.1 % of the dataset and it is impractical to set a dummy scorer name(that makes no sense, right?)
* Keep the rows where we have a <i>scorer</i> but <i>minute</i> is missing. Missing <i>minute</i>s can then be replaced by the mode of the <i>minute</i>s, i.e. the minute in which most goals are usually scored. This will preserve information as well as accuracy.

In [7]:
all_goals = matches.merge(right=goals, left_on=['date', 'home_team', 'away_team'], right_on=['date', 'home_team', 'away_team'], how='inner' )
all_goals

date  home_team away_team  home_score  away_score  \
0      1916-07-02      Chile   Uruguay           0           4   
1      1916-07-02      Chile   Uruguay           0           4   
2      1916-07-02      Chile   Uruguay           0           4   
3      1916-07-02      Chile   Uruguay           0           4   
4      1916-07-06  Argentina     Chile           6           1   
...           ...        ...       ...         ...         ...   
41003  2022-12-18  Argentina    France           3           3   
41004  2022-12-18  Argentina    France           3           3   
41005  2022-12-18  Argentina    France           3           3   
41006  2022-12-18  Argentina    France           3           3   
41007  2022-12-18  Argentina    France           3           3   

           tournament          city    country  neutral       team  \
0        Copa América  Buenos Aires  Argentina     True    Uruguay   
1        Copa América  Buenos Aires  Argentina     True    Uruguay   
2        Copa América  Buenos Aires  Argentina     True    Uruguay   
3        Copa América  Buenos Aires  Argentina     True    Uruguay   
4        Copa América  Buenos Aires  Argentina    False  Argentina   
...               ...           ...        ...      ...        ...   
41003  FIFA World Cup        Lusail      Qatar     True  Argentina   
41004  FIFA World Cup        Lusail      Qatar     True     France   
41005  FIFA World Cup        Lusail      Qatar     True     France   
41006  FIFA World Cup        Lusail      Qatar     True  Argentina   
41007  FIFA World Cup        Lusail      Qatar     True     France   

                 scorer  minute  own_goal  penalty  
0       José Piendibene    44.0     False    False  
1      Isabelino Gradín    55.0     False    False  
2      Isabelino Gradín    70.0     False    False  
3       José Piendibene    75.0     False    False  
4         Alberto Ohaco     2.0     False    False  
...                 ...     ...       ...      ...  
41003    Ángel Di María    36.0     False    False  
41004     Kylian Mbappé    80.0     False     True  
41005     Kylian Mbappé    81.0     False    False  
41006      Lionel Messi   109.0     False    False  
41007     Kylian Mbappé   118.0     False     True  

[41008 rows x 14 columns]

Since, the goals data in the given dataset for the years from 1872 to 1916 is not available but the match data is, I have decided not to delete the matches table because it still has some extra data that is nowhere else and might be useful later on.

In [9]:
matches.shape, goals.shape, all_goals.shape, penalties.shape

((44557, 9), (41008, 8), (41008, 14), (547, 4))